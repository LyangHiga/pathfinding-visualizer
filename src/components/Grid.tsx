import React from "react";

import Node from "./Node";

import { GridProps } from "./types/GridTypes";

function Grid(props: GridProps) {
  const {
    grid,
    isWeighted,
    handleMouseDown,
    handleMouseEnter,
    handleMouseUp,
  } = props;

  return (
    <div style={{ textAlign: "center" }}>
      {grid.grid.map((row, rowIdx) => {
        return (
          <div key={rowIdx}>
            {row.map((node) => {
              const { row, col, val, weight } = node;
              return (
                <Node
                  key={val}
                  val={val}
                  col={col}
                  row={row}
                  onMouseDown={handleMouseDown}
                  onMouseEnter={handleMouseEnter}
                  onMouseUp={handleMouseUp}
                  weight={weight}
                  isWeighted={isWeighted}
                ></Node>
              );
            })}
          </div>
        );
      })}
    </div>
  );
}

export default Grid;
