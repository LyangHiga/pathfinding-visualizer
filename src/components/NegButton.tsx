import React, { useState } from "react";
import {
  Button,
  Typography,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
  List,
  ListItem,
  ListItemText,
} from "@material-ui/core";

import bellmanFord from "../algorithms/bellmanFord";
import { valToIndx } from "../helpers/gridHelper";

import styles from "../styles/NavStyles";
import NegButtonProps from "./types/NegButtonTypes";

function NegButton(props: NegButtonProps) {
  const {
    grid,
    start,
    target,
    disable,
    btn,
    handleClick,
    setOpenDrawer,
    clear,
    newMaze,
    setIsWeighted,
  } = props;
  const classes = styles();
  const [rowStart, colStart] = valToIndx(start, grid.nCols);
  const [rowTarget, colTarget] = valToIndx(target, grid.nCols);

  const [openAlert, setOpenAlert] = useState(false);

  const handleClickOpenAlert = () => {
    setOpenAlert(true);
  };

  const handleCloseAlert = () => {
    setOpenAlert(false);
  };

  const bfDialog = (
    <Dialog
      open={openAlert}
      onClose={handleCloseAlert}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
    >
      <DialogTitle id="alert-dialog-title">
        How to run Bellman-Ford?
      </DialogTitle>
      <DialogContent>
        <DialogContentText id="alert-dialog-description">
          * Recommended Mode: Will Clear Grid, add to layers of walls and then
          run Bellman-Ford. PS: Put both start and finsih node in the same
          connected component is necessary for search min Path.
        </DialogContentText>
        <DialogContentText>
          * Free-Style: Won't change any grid property and you can run as you
          want.
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button
          onClick={async () => {
            setOpenDrawer(false);
            handleCloseAlert();
            clear();
            await newMaze();
            await newMaze();
            setIsWeighted(true);
            handleClick(
              bellmanFord(
                grid,
                grid.grid[rowStart][colStart],
                grid.grid[rowTarget][colTarget]
              )
            );
          }}
          color="primary"
        >
          Recommended
        </Button>
        <Button
          onClick={async () => {
            setOpenDrawer(false);
            handleCloseAlert();
            handleClick(
              bellmanFord(
                grid,
                grid.grid[rowStart][colStart],
                grid.grid[rowTarget][colTarget]
              )
            );
          }}
          color="primary"
          autoFocus
        >
          Free-Style
        </Button>
      </DialogActions>
    </Dialog>
  );

  const negBtn = (
    <div>
      <Button
        onClick={handleClickOpenAlert}
        disabled={disable}
        className={classes.button}
      >
        <Typography className={classes.text}>Bellman-Ford</Typography>
      </Button>
      {bfDialog}
    </div>
  );

  const negBtnList = (
    <List disablePadding>
      <ListItem
        onClick={handleClickOpenAlert}
        divider
        button
        className={classes.button}
        disabled={disable}
      >
        <ListItemText className={classes.listItemText}>
          Bellman-Ford
        </ListItemText>
      </ListItem>
      {bfDialog}
    </List>
  );

  return btn ? negBtn : negBtnList;
}

export default NegButton;
