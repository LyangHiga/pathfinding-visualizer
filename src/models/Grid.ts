import Node from "./Node";
import { createAdjList, getVal } from "../helpers/gridHelper";

export default class Grid {
  public grid: Node[][];
  public start: number;
  public target: number;
  public readonly nRows: number;
  public readonly nCols: number;
  // min and max weight allowed
  public readonly min: number;
  public readonly max: number;
  constructor(
    start: number,
    target: number,
    nRows: number,
    nCols: number,
    // min and max weight
    max = 1,
    min = 1
  ) {
    this.grid = [];
    this.start = start;
    this.target = target;
    this.nCols = nCols;
    this.nRows = nRows;
    this.min = min;
    this.max = max;
    for (let row = 0; row < nRows; row++) {
      const currentRow: Node[] = [];
      for (let col = 0; col < nCols; col++) {
        // get random weight for all edges of this node
        const weight = Math.floor(Math.random() * (max - min) + min);
        const val = getVal(row, nCols, col);
        const AdjList = createAdjList(val, col, row, nRows, nCols);
        const newNode = new Node(col, row, val, start, target, AdjList, weight);
        currentRow.push(newNode);
      }
      this.grid.push(currentRow);
    }
  }
}
