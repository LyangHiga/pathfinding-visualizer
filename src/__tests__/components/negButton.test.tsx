import React from "react";
import { shallow } from "enzyme";
import NegButton from "../../components/NegButton";
import Grid from "../../models/Grid";

const START = 10;
const TARGET = 14;

const NCOLS = 5;
const NROWS = 5;

//       0  1   2   3   4
//       5  6   7   8   9
// g =   S  11  12  13  T
//      15  16  17  18  19
//      20  21  22  23  24
describe("Neg Button", () => {
  it("renders correctly", () => {
    const grid = new Grid(START, TARGET, NROWS, NCOLS);
    const wrapper = shallow(
      <NegButton
        grid={grid}
        disable={true}
        btn={true}
        handleClick={jest.fn()}
        setOpenDrawer={jest.fn()}
        clear={jest.fn()}
        newMaze={jest.fn()}
        setIsWeighted={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
    // On the first run of this test, Jest will generate a snapshot file automatically.
  });
});

interface Element {
  key: string;
  props: {
    disabled: boolean;
    onClick: () => void;
    children?: {
      props?: {
        children?: string;
      };
    };
  };
}

describe("Neg Button", () => {
  it("Random size Grid with disable prop false and btn style, should render enable neg btn", () => {
    // Random size grid
    const RANDOM_ROWS = Math.floor(Math.random() * 10) + 1;
    const RANDOM_COLS = Math.floor(Math.random() * 20) + 2;
    // start = 0, target = 1 are guaranteed to exist
    const g = new Grid(0, 1, RANDOM_ROWS, RANDOM_COLS);
    const wrapper = shallow(
      <NegButton
        grid={g}
        disable={false}
        btn={true}
        handleClick={jest.fn()}
        setOpenDrawer={jest.fn()}
        clear={jest.fn()}
        newMaze={jest.fn()}
        setIsWeighted={jest.fn()}
      />
    );
    const btns = wrapper.props().children;
    const negBtn: Element = btns.filter((e: Element) => e.key === "neg-btn")[0];

    expect(negBtn).not.toBeUndefined();
    expect(negBtn).not.toBeNull();
    // typography inside btn named as "Bellman-Ford"
    expect(negBtn.props.children!.props!.children).toBe("Bellman-Ford");
    // onClick func that opens dialog
    expect(negBtn.props.onClick).not.toBeUndefined();
    // should be enable
    expect(negBtn.props.disabled).toBeFalsy();
  });
});
