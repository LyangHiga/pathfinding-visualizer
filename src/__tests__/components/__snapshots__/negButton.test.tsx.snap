// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-button-4"
    disabled={true}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-text-7"
    >
      Bellman-Ford
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Dialog))
    aria-describedby="alert-dialog-description"
    aria-labelledby="alert-dialog-title"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="alert-dialog-title"
    >
      How to run Bellman-Ford?
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))
        id="alert-dialog-description"
      >
        * Recommended Mode: Will Clear Grid, add to layers of walls and then run Bellman-Ford. PS: Put both start and finsih node in the same connected component is necessary for search min Path.
      </WithStyles(ForwardRef(DialogContentText))>
      <WithStyles(ForwardRef(DialogContentText))>
        * Free-Style: Won't change any grid property and you can run as you want.
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
      >
        Recommended
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        autoFocus={true}
        color="primary"
        onClick={[Function]}
      >
        Free-Style
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
